### ISSUE TITLE
Add mapZIOPar and mapZIOParUnordered operations to ZPipeline

### ISSUE THREAD

**Issue by pwliwanow** [zio/zio#7715]
Thanks for ZIO; it's a great library! 
It would be useful (it would certainly be useful to me, as I am migrating from Akka Streams to ZIO Streams, but I am sure others are in the same boat nowadays) to expose `mapZIOPar` and `mapZIOParUnordered` on `ZPipeline`.

I started working on a PR to add that, but it turned out to be more complex than I had hoped. So below I'll share some details in case it's useful for someone that would like to add these operators.
I added the `mapZIOPar` to `ZPipeline`:
```scala
def mapZIOPar[Env, Err, In, Out](n: => Int)(f: In => ZIO[Env, Err, Out])(implicit
    trace: Trace
  ): ZPipeline[Env, Err, In, Out] =
    new ZPipeline(
      ZChannel
        .identity[Nothing, Chunk[In], Any]
        .concatMap(ZChannel.writeChunk(_))
        .mapOutZIOPar[Env, Err, Out](n)(f)
        .mapOut(Chunk.single)
    )
```
However, updating `mapZIOPar` in `ZStream` to::
```scala
def mapZIOPar[R1 <: R, E1 >: E, A2](n: => Int)(f: A => ZIO[R1, E1, A2])(implicit
    trace: Trace
  ): ZStream[R1, E1, A2] =
   self >>> ZPipeline.mapZIOPar(n)(f)
```
causes the `mapZIOPar` tests in `ZStream` to fail with "Unexpected end of input for channel execution".

To narrow down which operation is at fault, I expanded the new `mapZIOPar` implementation in `ZStream`. 
Here's a version for which `mapZIOPar` tests succeed:
```scala
  def mapZIOPar[R1 <: R, E1 >: E, A2](n: => Int)(f: A => ZIO[R1, E1, A2])(implicit
    trace: Trace
  ): ZStream[R1, E1, A2] =
    new ZStream(
      channel
        .concatMap(ZChannel.writeChunk(_))
        .mapOutZIOPar[R1, E1, A2](n)(f)
        .pipeTo(ZChannel.identity[E1, A2, Any].mapOut(Chunk.single))
    )
```
And here's a version for which `mapZIOPar` tests fail:
```scala
  def mapZIOPar[R1 <: R, E1 >: E, A2](n: => Int)(f: A => ZIO[R1, E1, A2])(implicit
    trace: Trace
  ): ZStream[R1, E1, A2] =
    new ZStream(
      channel
        .concatMap(ZChannel.writeChunk(_))
        .pipeTo(ZChannel.identity[E, A, Any].mapOutZIOPar[R1, E1, A2](n)(f).mapOut(Chunk.single))
    )
```
In summary, after moving `mapOutZIOPar` inside `pipeTo` (which is kind of what happens when the impl would look like `self >>> ZPipeline.mapZIOPar(n)(f)`), tests fail with the failure mentioned earlier.
